cmake/introspection.cmake

```diff
@@ -0,0 +1,255 @@
+# Fail if neither are available.
+check_cxx_source_compiles("
+#include <ctime>
+
+int main()
+{
+  gmtime_r((const time_t*)nullptr, (struct tm*)nullptr);
+}
+" HAVE_GMTIME_R)
+if(NOT HAVE_GMTIME_R)
+  check_cxx_source_compiles("
+  #include <ctime>
+
+  int main()
+  {
+    gmtime_s((struct tm*)nullptr, (const time_t*)nullptr);
+  }
+  " HAVE_GMTIME_S)
+  if(NOT HAVE_GMTIME_S)
+    message(FATAL_ERROR "Both gmtime_r and gmtime_s are unavailable.")
+  endif()
+endif()
+
+check_cxx_source_compiles("
+#include <cstdlib>
+
+int main()
+{
+  return std::system(nullptr);
+}
+" HAVE_STD_SYSTEM)
+check_cxx_source_compiles("
+#include <stdlib.h>
+
+int main()
+{
+  return ::_wsystem(NULL);
+}
+" HAVE__WSYSTEM)
+if(HAVE_STD_SYSTEM OR HAVE__WSYSTEM)
+  set(HAVE_SYSTEM 1)
+endif()
+
+CHECK_INCLUDE_FILE_CXX(string.h HAVE_STRING_H)
+if(HAVE_STRING_H)
+  check_cxx_source_compiles("
+  #include <string.h>
+
+  int main()
+  {
+    char buf[100];
+    char x = *strerror_r(0, buf, sizeof buf);
+    char* p = strerror_r(0, buf, sizeof buf);
+    return !p || x;
+  }
+  " STRERROR_R_CHAR_P)
+endif()
+
+# Check for malloc_info (for memory statistics information in getmemoryinfo).
+check_cxx_symbol_exists(malloc_info "malloc.h" HAVE_MALLOC_INFO)
+
+# Check for mallopt(M_ARENA_MAX) (to set glibc arenas).
+check_cxx_source_compiles("
+#include <malloc.h>
+
+int main()
+{
+  mallopt(M_ARENA_MAX, 1);
+}
+" HAVE_MALLOPT_ARENA_MAX)
+
+# Check for posix_fallocate().
+check_cxx_source_compiles("
+#ifdef __linux__
+#ifdef _POSIX_C_SOURCE
+#undef _POSIX_C_SOURCE
+#endif
+#define _POSIX_C_SOURCE 200112L
+#endif // __linux__
+#include <fcntl.h>
+
+int main()
+{
+  return posix_fallocate(0, 0, 0);
+}
+" HAVE_POSIX_FALLOCATE)
+
+# Check for strong getauxval() support in the system headers.
+check_cxx_source_compiles("
+#include <sys/auxv.h>
+
+int main()
+{
+  getauxval(AT_HWCAP);
+}
+" HAVE_STRONG_GETAUXVAL)
+
+# Check for different ways of gathering OS randomness:
+# - Linux getrandom()
+check_cxx_source_compiles("
+#include <unistd.h>
+#include <sys/syscall.h>
+#include <linux/random.h>
+
+int main()
+{
+  syscall(SYS_getrandom, nullptr, 32, 0);
+}
+" HAVE_SYS_GETRANDOM)
+
+# - BSD getentropy()
+check_cxx_symbol_exists(getentropy "unistd.h;sys/random.h" HAVE_GETENTROPY_RAND)
+
+# - BSD sysctl()
+check_cxx_source_compiles("
+#include <sys/types.h>
+#include <sys/sysctl.h>
+
+#ifdef __linux__
+#error \"Don't use sysctl on Linux, it's deprecated even when it works\"
+#endif
+
+int main()
+{
+  sysctl(nullptr, 2, nullptr, nullptr, nullptr, 0);
+}
+" HAVE_SYSCTL)
+
+# - BSD sysctl(KERN_ARND)
+check_cxx_source_compiles("
+#include <sys/types.h>
+#include <sys/sysctl.h>
+
+#ifdef __linux__
+#error \"Don't use sysctl on Linux, it's deprecated even when it works\"
+#endif
+
+int main()
+{
+  static int name[2] = {CTL_KERN, KERN_ARND};
+  sysctl(name, 2, nullptr, nullptr, nullptr, 0);
+}
+" HAVE_SYSCTL_ARND)
+
+check_cxx_source_compiles("
+#include <stdint.h>
```

Comment:
Any reason we can't use the c++ headers here?

Reply:
Thanks! [Fixed](https://github.com/hebasto/bitcoin/pull/6#issuecomment-1436835797).